pyvisa_py/tcpip.py

```diff
@@ -64,10 +64,11 @@ def __new__(
         return newcls(resource_manager_session, resource_name, parsed, open_timeout)
 
 
-@Session.register(constants.InterfaceType.tcpip, "HISLIP")
 class TCPIPInstrHiSLIP(Session):
     """A TCPIP Session built on socket standard library using HiSLIP protocol."""
 
+    session_type = (constants.InterfaceType.tcpip, "HISLIP")
```

Comment:
From the VISA point of view the resource_class is INSTR I think

Reply:
Doesn't the session_type value need to be unique so that the attribute machinery will work?  The VISA attributes for HiSLIP aren't the same as those for VXI-11.  If we use INSTR for both VXI-11 and HiSLIP, wouldn't the HiSLIP attributes  appear on VXI-11 instances, where they wouldn't make any sense?

(the following is from https://documentation.help/VISA.NET/Using%20the%20TCPIP%20Interface.htm)

The HiSLIP protocol introduces three new VISA attributes for HiSLIP devices:

VI_ATTR_TCPIP_HISLIP_OVERLAP_EN  -  If VI_TRUE, enables overlap mode for HiSLIP. If VI_FALSE, the synchronous mode is used. Refer to [Synchronous and Overlap Modes](https://documentation.help/VISA.NET/Using%20the%20TCPIP%20Interface.htm#RTF320033003200300037003a00) for a description of these modes.
VI_ATTR_TCPIP_HISLIP_VERSION  -  Returns the ViVersion in use by the current HiSLIP session. For example, HiSLIP version 1.0 returns  a ViVersion value of 0x00100000.
VI_ATTR_TCPIP_HISLIP_MAX_MESSAGE_KB  -  Specifies the largest size HiSLIP message VISA will accept in units of kilobytes;  defaults to 1024 (a 1 MB maximum message size).

Reply:
If you look at the implementation of attributes (https://github.com/pyvisa/pyvisa/blob/main/pyvisa/attributes.py) and the default behavior of sessions (https://github.com/pyvisa/pyvisa-py/blob/main/pyvisa_py/sessions.py#L620) you will see that:
- we only use valid resource class 
- their use simply allow to quickly discard unsupported attributes
- even if the test success we may still return StatusCode.error_nonsupported_attribute

As a consequence I still believe we should use INSTR for both VXI-11 and HiSLIP and simply provide different implementation for supported attributes.  

BTW do you plan to add support for any attribute as part of this PR ?

Reply:
According to rule 5.1.16 (https://www.ivifoundation.org/docs/vpp43_2016-02-26.pdf), it's mandatory to support VI_ATTR_TCPIP_ADDR, VI_ATTR_TCPIP_HOSTNAME, VI_ATTR_TCPIP_IS_HISLIP, and VI_ATTR_TCPIP_DEVICE_NAME.

And according to rule 5.1.17, HiSLIP implementations must support VI_ATTR_TCPIP_PORT, VI_ATTR_TCPIP_NODELAY, VI_ATTR_TCPIP_KEEPALIVE, VI_ATTR_TCPIP_HISLIP_OVERLAP_EN, VI_ATTR_TCPIP_HISLIP_VERSION, and VI_ATTR_TCPIP_HISLIP_MAX_MESSAGE_KB.

I could use some pointers on how to implement these attributes.

I see another applicable rule, 5.1.31, which says that if no device name is specified, a VXI11 connection should be attempted, and if that fails, a HiSLIP connection should be attempted.  I'll have to think about how to implement that one.  Maybe that won't happen any time soon.

Reply:
I will look into the attributes. But I may not block your PR on it. PyVISA-py implement very few of them and I do not want to block your work on this. However a number of those are static values and should be easy.

Reply:
I added get/set for keepalive and max_message_kb for HiSLIP.  While I was adding the keepalive, I noticed that the keepalive for VXI11 is Linux specific...it breaks on OSX (and presumably on Windows, but I haven't tried that).

See https://stackoverflow.com/questions/12248132/how-to-change-tcp-keepalive-timer-using-python-script